### ISSUE TITLE
IRM implementation

### ISSUE THREAD

**Issue by yelobean** [facebookresearch/DomainBed#85]
I wonder about your IRM implementation (below).
```
    def _irm_penalty(logits, y):
        device = "cuda" if logits[0][0].is_cuda else "cpu"
        scale = torch.tensor(1.).to(device).requires_grad_()
        loss_1 = F.cross_entropy(logits[::2] * scale, y[::2])
        loss_2 = F.cross_entropy(logits[1::2] * scale, y[1::2])
        grad_1 = autograd.grad(loss_1, [scale], create_graph=True)[0]
        grad_2 = autograd.grad(loss_2, [scale], create_graph=True)[0]
        result = torch.sum(grad_1 * grad_2)
        return result
```

IRM paper indicates that penalty term is gradient's square.
<img width="395" alt="image" src="https://user-images.githubusercontent.com/32060985/148762328-23e93090-fef3-43a7-9fdc-72764d2ae3ef.png">

Also, IRM penalty from 'https://github.com/facebookresearch/InvariantRiskMinimization' is the below code.
```
  def penalty(logits, y):
    scale = torch.tensor(1.).cuda().requires_grad_()
    loss = mean_nll(logits * scale, y)
    grad = autograd.grad(loss, [scale], create_graph=True)[0]
    return torch.sum(grad**2)
```

I do not think your implementation is wrong because you are the authors of the IRM paper.

However, ... 
Could you teach me why you exploit the dot product between gradient_odd and gradient_even instead of gradient's square?
Could you please explain if I have misunderstood something?

**Comment by igul222** [comment]
It's not a bug; see section 3.2 of the IRM paper. In short, splitting the minibatch like this gives an unbiased estimator of the full-dataset squared gradient norm; squaring the minibatch gradient would be biased.

**Comment by yelobean** [comment]
I understand!
Thanks for your kind explanation.
